### DISCUSSION TITLE
Reading data from html tables

### DISCUSSION THREAD

**Discussion by onnimonni** [duckdb/duckdb#13764] [Ideas]
Duckdb has almost completely replaced Excel and Google sheets for me. It's great that I can do so many ad-hoc queries directly from command line to read json and csv. SQL provides much better ergonomics than for example using `jq`.

One feature I'm still missing is the ability to read data from html tables. I can easily extract information from Wikipedia or other websites using the `=IMPORTHTML()` function in Google sheets. See more examples here: http://www.statsmapsnpix.com/2022/04/how-to-import-wikipedia-table-directly.html?m=1

I tried to find other command line tools to pipe things into duckdb stdin but they were pretty hard to use.

Would someone know how to do this with current duckdb or could I suggest this as a new feature?

**Comment by yoasty** [comment]
Is it maybe possible to use the HTTP_client extension to read tables from pages?  I really would appreciate getting tables from some sources. worldodometers, wikipedia. etc. 

https://duckdb.org/community_extensions/extensions/http_client.html

**Reply by EED85** [reply]
I would love that Feature as well!

**Reply by onnimonni** [reply]
The `webbed` extension almost works for this: https://github.com/teaguesterling/duckdb_webbed/issues/5

My problem is that it creates arrays of varchars instead of actually reading the column names. Maybe someone here is able to figure out how to use it?

**Comment by onnimonni** [comment]
It's pretty complex and hacky way to it but this is possible with [duckdb webbed extension](https://github.com/teaguesterling/duckdb_webbed) like this:

```sql
INSTALL webbed FROM community;
LOAD webbed;

-- FIXME: Change these 2 variables for your website
SET VARIABLE html_content = (
    SELECT html::VARCHAR
    FROM read_html('https://en.wikipedia.org/wiki/List_of_most_expensive_buildings')
    LIMIT 1
);
SET VARIABLE html_table_index = 2;

WITH html_data AS (
    FROM html_extract_tables(getvariable('html_content'))
    WHERE table_index = getvariable('html_table_index')
), headers AS (
    SELECT
        *,
        ROW_NUMBER() OVER(ORDER BY TRUE) AS column_number,
    FROM (
        SELECT UNNEST(columns) AS column_name,
        FROM html_data
        WHERE row_index = 0
    )
), values AS (
    SELECT
        *,
        ROW_NUMBER() OVER(PARTITION BY row_index ORDER BY TRUE) AS column_number
    FROM (
        SELECT row_index, UNNEST(columns) AS column_value,
        FROM html_data
        WHERE row_index > 0
    )
), columns_with_values AS (
    SELECT * EXCLUDE(column_number)
    FROM values
    JOIN headers USING(column_number)
), data AS (
    SELECT * EXCLUDE(row_index) FROM (
        PIVOT columns_with_values ON column_name
        USING MAX(column_value)
        ORDER BY row_index ASC
    )
)
FROM data;
```

**Reply by onnimonni** [reply]
I hope someone can improve the webbed extension or my hacky examples. It's pretty problematic to select table based on the index.

For example it's pretty problematic to select the html table based on index in wikipedia because of this:

<img width="1235" height="736" alt="Screenshot 2025-09-06 at 12 52 03" src="https://github.com/user-attachments/assets/9c9f41a9-9bd8-4289-852f-01f2bbceacd6" />

It would be much better to select based on html `id` or `class`. For example the table in my example looks like this:
```html
<table class="wikitable sortable jquery-tablesorter" border="1">
...
</table>
```

So selecting the a table with `class=wikitable` is much more preferred instead of the table index.

**Reply by MarkMellink-RiffOnline** [reply]
Does this still work for you? If I execute this now I get `SQL Error: Binder Error: Referenced column "html" not found in FROM clause!
Candidate bindings: "title", "meta"`

**Reply by onnimonni** [reply]
I haven't tried in months anymore. It's so trivial to build new extension to Duckdb nowadays with Claude Code so I imagine you can just ask it to do a community extension which does this.

**Comment by onnimonni** [comment]
I made a new version for the [duckdb-crawler](https://github.com/midwork-finds-jobs/duckdb-crawler) extension I have been working with lately: https://github.com/duckdb/community-extensions/pull/1113.

It was trivial to do with Claude Code.

After the PR is merged you should be able to do following which is much better than what `=HTMLIMPORT()` ever was.

@MarkMellink-RiffOnline these are typically merged on the next morning so then you can use it.

Use it responsibly 🙇 

```sql
-- Install the crawler extension
INSTALL crawler FROM community;
LOAD crawler;

-- Identify yourself to the web servers you are using, by default this reads also the /robots.txt and will not crawl the page if it's not allowed
SET crawler_user_agent = 'MyBot/1.0 (+https://example.com/bot)';

-- Get a table from the html page
FROM read_html('https://en.wikipedia.org/wiki/List_of_most_expensive_buildings', '#bodyContent table.wikitable.sortable');

-- Get first match of unordered/ordered list from the html page
FROM read_html('https://wikipedia.org/wiki/List_of_programming_languages', 'div.div-col ul', 1);

-- Parse the the javascript to AST in the <script> tags and extract the value of simple javascript variables without evaluating the code
-- This works only with simple data structures for foreseeable future because I don't intend to add a real browser in the crawler extension.
FROM read_html('https://en.wikipedia.org/wiki/List_of_most_expensive_buildings', 'js=RLCONF');
```

**Reply by MarkMellink-Koalitix** [reply]
Thanks I'll give this a go once it's available. I'm actually attending the DuckDB Extension Development Workshop in Amsterdam during the dev days next week, so after I might give extension building a go myself!

**Reply by onnimonni** [reply]
If you can contribute to the crawler directly it would be great so then we wouldn't need to build million overlapping systems but I do understand if you would want to have something smaller since this takes into account a lot of stuff.

**Reply by MarkMellink-Koalitix** [reply]
No definitely, I didn't mean building this exact functionality as an extension, more that I'll be trying it for the first time! But if there's functionality I think makes sense as an addition I'll try and contribute.

**Comment by EED85** [comment]
Hi @onnimonni,

your extensions looks very promising! 
I am using duckdb `1.4.3` aon windows, installed via `winget`.

After successful installation and loading, duckdb does not recognize the `read_html` function. What am i doing wrong?

```
DuckDB v1.4.3 (Andium) d1dc88f950
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D -- Install the crawler extension
D INSTALL crawler FROM community;
D LOAD crawler;
D
D -- Identify yourself to the web servers you are using, by default this reads also the /robots.txt and will not crawl the page if it's not allowed
D SET crawler_user_agent = 'MyBot/1.0 (+https://example.com/bot)';
D
D -- Get a table from the html page
D FROM read_html('https://en.wikipedia.org/wiki/List_of_most_expensive_buildings', '#bodyContent table.wikitable.sortable');
Catalog Error:
Table Function with name read_html does not exist!
Did you mean "read_blob"?

LINE 1: FROM read_html('https://en.wikipedia.org/wiki/List_of_most_expen...
```

i guess i need to wait for the PR to be merged?

**Reply by onnimonni** [reply]
Sorry! It seems that I forgot to tag that new feature to the latest release.

> The read_html function was added on 2026-01-22 but v0.1.0 release was 2026-01-18. The community
  extension doesn't have read_html yet - you need to cut a new release.
 
 I will make a new release soon and hopefully it gets merged on monday. In the mean time feel free to git clone and build the extension locally.

**Reply by onnimonni** [reply]
Hey @EED85.

The fix is merged now.

Run the command below first and then the examples above should work. I tested the commands above locally (on MacOS) and it worked now just fine.
```
# Update crawler and other extensions
$ duckdb -c "UPDATE EXTENSIONS"
```

**Reply by EED85** [reply]
Great, the `read_html` Function now works with your given example!

FYI: Updating extensions via `Powershell` on Windows did not work:

```ps1
PS C:\Users\xx> duckdb -c "UPDATE EXTENSIONS"
IO Error:
Extension updating failed when trying to install 'crawler', original error: Failed to delete file "C:\Users\xx\.duckdb\extensions\v1.4.3\windows_amd64\crawler.duckdb_extension": Der Prozess kann nicht auf die Datei zugreifen, da sie von einem anderen Prozess verwendet wird.
```

Even after a restart, the same error occured. I deleted the manually via File Explorer and installed the crawler Extension again.

**Comment by EED85** [comment]
Just a comment here:
I can use this great extension now privately, but at my office, the duckdb extensions are blocked via Firewall.

Its a pitty (for me), that the extensions are not installable via conda or pip, as we are using a mirrored PyPi and Conda Artifactory.

See https://github.com/duckdb/duckdb-python/discussions/39
